src/main/java/uk/sky/cirrus/locking/Lock.java

```diff
@@ -0,0 +1,157 @@
+    private final UUID client;
+    private boolean released;
+
+    private Lock(String name, Session session, UUID client) {
+        this.name = name;
+        this.session = session;
+        this.client = client;
+    }
+
+    /**
+     * @param clientId   Identifier for the owner of the lock
+     * @param lockConfig {@code LockConfig} for configuring the lock to migrate the schema
+     * @param keyspace   Name of the keyspace which is to be used for migration
+     * @param session    Cassandra {@code Session} to use while acquiring the lock
+     * @return the {@code Lock} object
+     * @throws CannotAcquireLockException if instance cannot acquire lock within the specified time interval or execution of query to insert lock fails
+     */
+    static Lock acquire(LockConfig lockConfig, String keyspace, Session session, UUID clientId) {
+
+        final String clientName = keyspace + ".schema_migration";
+        int acquireAttempts = 1;
+
+        log.info("Attempting to acquire lock for '{}', using client id '{}'", clientName, clientId);
+        long startTime = System.currentTimeMillis();
+
+        final Statement query = new SimpleStatement("INSERT INTO locks.locks (name, client) VALUES (?, ?) IF NOT EXISTS", clientName, clientId);
+
+        while (true) {
+            try {
+                final ResultSet resultSet = tryAcquire(session, clientId, query);
+                final Row currentLock = resultSet.one();
+                boolean hasAcquiredLock = currentLock.getBool("[applied]");
+
+                if (hasAcquiredLock) {
+                    log.info("Lock acquired for '{}' by client '{}' after {} attempts", clientName, clientId, acquireAttempts);
+                    return new Lock(clientName, session, clientId);
+                } else if (clientId.equals(currentLock.getUUID("client"))) {
+                    return new Lock(clientName, session, clientId);
+                } else {
+                    waitToAcquire(lockConfig, clientId, acquireAttempts, startTime, currentLock.getUUID("client"));
+                }
+
+            } catch (WriteTimeoutException wte) {
+                log.warn("Query to acquire lock for {} failed to execute: {}", clientId, wte.getMessage());
+
+            }
+            acquireAttempts++;
+        }
+    }
+
+    /**
+     * @throws CannotReleaseLockException if execution of query to remove lock fails
+     */
+    void release() {
+        final Statement query = new SimpleStatement("DELETE FROM locks.locks WHERE name = ? IF client = ?", name, client);
+
+        while(!this.released){
+            try {
+                ResultSet resultSet = session.execute(query);
+                Row result = resultSet.isExhausted() ? null : resultSet.one();
+                this.released = handleLockRelease(result);
+            } catch (WriteTimeoutException wte) {
+                log.warn("Query to release lock for {} failed to execute: {}", client, wte.getMessage());
```

Comment:
Back off before trying again, don't want to be in a tight loop.

Also if someone else has the lock we can consider it released from a previous right time out as well?